EducationalAssignments/UndoPartOne.md

```diff
@@ -0,0 +1,303 @@
+occurred:
+
+ * using `print` instead of `log`:
+
+Repy is a subset of Python, but its syntax is slightly different.  For example,
+Python's `print` statement cannot be used; Repy has `log` for that. For a full
+list of acceptable syntax please see
+[https://github.com/SeattleTestbed/docs/blob/master/Programming/RepyV2API.md]
+
+ * command line errors:
+
+**files are missing:** In the above command line call, you must have `repy.py`,
+restrictions.default, encasementlib.r2py, the security layer and the program you
+want to run in the current working directory.  If any or all of the above files
+are not in that directory then you will not be able to run repy files.  
+
+<!--
+AR: This doesn't apply when building from source or getting the runtime tarball only (it does for clearinghouse downloads).
+ * Downloading the wrong version of Seattle:
+
+Seattle is operating system dependent.  If you download the Windows
+version, you need to use the Windows command line.  For Windows 7 this is
+PowerShell.  You can open a new terminal by going to start, search, type
+powershell.  If you downloaded the Linux version you must use a Linux OS
+and Linux terminal.  
+
+-->
+
+
+### Tutorials for Repy and Python
+
+Now that you have Repy and Python, you may need a refresher on how to use them.
+The following tutorials provide this information.
+
+ * Official [Python tutorial](http://docs.python.org/tutorial/)
+ * [Differences between RepyV2 and Python](../Programming/PythonVsRepyV2.md)
+ * List of [RepyV2 API calls](../Programming/RepyV2API.md)
+
+
+
+## Building the security layer
+
+The following program is a sample security layer, it is not complete and does
+not handle all cases required by the API. Remember, you have no idea how the
+attacker will try to penetrate your security layer, so it is important that you
+leave nothing to chance!  
+
+
+### A basic (and inadequate) defense
+
+Time to start coding!  Let's inspect a basic security layer.  
+
+```py
+"""
+This security layer inadequately handles the undo functionality
+
+Note:
+    This security layer uses encasementlib.r2py, restrictions.default, repy.py and Python
+    Also you need to give it an application to run.
+    python repy.py restrictions.default encasementlib.r2py [security_layer].r2py [attack_program].r2py 
+    
+"""
+TYPE = "type"
+ARGS = "args"
+RETURN = "return"
+EXCP = "exceptions"
+TARGET = "target"
+FUNC = "func"
+OBJC = "objc"
+
+
+class LPFile():
+    def __init__(self,filename,create):
+        # globals
+        mycontext['debug'] = False   
+        self.LPfile = openfile(filename,create)
+        self.pending_data = ""
+        self.pending_offset = 0
+
+    def readat(self, bytes, offset):
+        # Read from the file using the sandbox's readat...
+        return self.LPfile.readat(bytes, offset)
+
+    def writeat(self,data,offset):
+        self.LPfile.writeat(data, offset)
+        self.pending_data += data
+        self.pending_offset = offset + len(data)
```

Comment:
This is broken in weird ways.  Let's give them an example that is wrong, but not for odd reasons.  For example, perhaps undo doesn't check if there was any pending data to discard and would have an exception in that case or undo fails to remove the buffer for writeat.

Reply:
Prof @JustinCappos,

> This is broken in weird ways. Let's give them an example that is wrong, but not for odd reasons. 

I got it.

I updated the code. How does it look to you?

Please let me know if you need more changes.